pydantic_ai_slim/pydantic_ai/models/adaptive.py

```diff
@@ -0,0 +1,297 @@
+from __future__ import annotations as _annotations
+
+import inspect
+import time
+from collections.abc import AsyncIterator, Awaitable, Callable, Sequence
+from contextlib import AsyncExitStack, asynccontextmanager, suppress
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Generic, TypeVar
+
+from opentelemetry.trace import get_current_span
+
+from pydantic_ai._run_context import RunContext
+from pydantic_ai.models.instrumented import InstrumentedModel
+
+from ..exceptions import FallbackExceptionGroup
+from ..settings import merge_model_settings
+from . import Model, ModelRequestParameters, StreamedResponse
+
+if TYPE_CHECKING:
+    from ..messages import ModelMessage, ModelResponse
+    from ..settings import ModelSettings
+
+AgentDepsT = TypeVar('AgentDepsT')
+
+
+@dataclass
+class AttemptResult:
+    """Record of a single attempt to use a model."""
+
+    model: Model
+    """The model that was attempted."""
+
+    exception: Exception | None
+    """The exception raised by the model, if any."""
+
+    timestamp: float
+    """Unix timestamp when the attempt was made."""
+
+    duration: float
+    """Duration of the attempt in seconds."""
+
+
+@dataclass
+class AdaptiveContext(Generic[AgentDepsT]):
+    """Context provided to the selector function."""
+
+    run_context: RunContext[AgentDepsT] | None
+    """Access to agent dependencies. May be None for non-streaming requests."""
+
+    models: Sequence[Model]
+    """Available models to choose from."""
+
+    attempts: list[AttemptResult]
+    """History of attempts in this request."""
+
+    attempt_number: int
+    """Current attempt number (1-indexed)."""
```

Comment:
Wouldn't this always be `len(attempts) + 1`? In that case it could be a `property`, or be omitted entirely